### PR TITLE
Louis/podcidrs bgp

### PR DESCRIPTION
**Pull request by ldelossa**
This pull adds the necessary commits to support BGP route withdrawal on a Node Delete event. 

After a Node Delete event is encountered an empty Advertisement will be plumbed to the BGP controller running inside the Cilium agent. 

After the event is received by our harness, we will flip a atomic to then shut the speaker down. When the BGP speaker is shut down it will no longer ingest concurrent events nor process any events that are still in its event queue.

### CODE REVIEW

File: pkg/bgp/speaker/events.go
Comment (christarazi) [inline pkg/bgp/speaker/events.go]: This doesn't need to be a pointer because `bool` is hashable. Making these fields pointers is only required for types that are not hashable such as maps or slices. 

I guess 
>// The following fields must be a pointers because they are not hashable
	// (read comparable) in Go.

was a bit misleading since it can be interpreted as all fields? Probably worth clarifying it.

File: pkg/bgp/speaker/events.go
Comment (christarazi) [inline pkg/bgp/speaker/events.go:54]: This change was probably supposed to be in the previous commit?

File: pkg/bgp/speaker/pod_cidr.go
Comment (christarazi) [inline pkg/bgp/speaker/pod_cidr.go]: `log.WithError(err) ...`

File: pkg/bgp/speaker/speaker.go
Comment (christarazi) [inline pkg/bgp/speaker/speaker.go]: Probably worth extracting out this error into its own variable via `errors.New()`

```
var (
  errShuttingDown = errors.New(...)
)
```

File: pkg/bgp/speaker/events.go
Comment (christarazi) [inline pkg/bgp/speaker/events.go]: Would be nice to reference the RFC here

File: pkg/bgp/speaker/events.go
Comment (ldelossa) [reply pkg/bgp/speaker/events.go]: You are correct. 

I'm not exactly aware of why this consideration is here in the first place? I didn't see any code doing comparisons of these event types and fields. Can you clue me in?

File: pkg/bgp/speaker/speaker.go
Comment (ldelossa) [reply pkg/bgp/speaker/speaker.go]: I like that :+1:

File: pkg/bgp/speaker/events.go
Comment (christarazi) [reply pkg/bgp/speaker/events.go]: The issue is that the event queuing uses a `map` under the hood, I believe. So if you were to remove all the pointers from the `nodeEvent`, you'll get a panic saying "unhashable type" or something like that when a `nodeEvent` comes in. https://blog.golang.org/maps#TOC_5.

Comment (ldelossa) [reply pkg/bgp/speaker/events.go]: got it. It's required by the k8s workqueue.

File: pkg/bgp/speaker/events.go
Comment (ldelossa) [reply pkg/bgp/speaker/events.go]: agreed.

**Comment by ldelossa** [comment]
@christarazi - handled your PR comments  - prob needs one last look over and if this is approved I'll merge right into your topic.

File: pkg/bgp/speaker/pod_cidr.go
Comment (christarazi) [inline pkg/bgp/speaker/pod_cidr.go]:
```suggestion
				log.WithError(err).Error("Failed to withdraw BGP routes")
```

File: pkg/bgp/speaker/speaker.go
Comment (christarazi) [inline pkg/bgp/speaker/speaker.go:232]:
```suggestion
// instructing it to withdraw all previously advertised
```

File: pkg/bgp/speaker/speaker.go
Comment (christarazi) [inline pkg/bgp/speaker/speaker.go]: I thought that I refactored the other subscriber implementations to return errors as well, but I'm only seeing the node subscriber methods returning errors? If the others haven't been implemented, I think it's worth doing a refactor commit to do that. Doesn't have to be part of this specific PR though.

File: pkg/bgp/speaker/speaker.go
Comment (christarazi) [inline pkg/bgp/speaker/speaker.go:40]: Does this need to be exported?

**Review by christarazi — APPROVED** [review]
Minor nits

File: pkg/bgp/speaker/speaker.go
Comment (ldelossa) [reply pkg/bgp/speaker/speaker.go]: I'll do this in this PR.

File: pkg/bgp/speaker/speaker.go
Comment (ldelossa) [reply pkg/bgp/speaker/speaker.go:40]: It would allow the k8s watcher to determine if it wants to react to the bgp speaker being shut down.

**Comment by ldelossa** [comment]
@christarazi - latest nits fixed, also new commit which addresses returning errors on all the event handling methods.

**Review by christarazi — APPROVED** [review]

**Comment by joestringer** [comment]
@christarazi @ldelossa Please be sure to check the labels like `dont-merge/*` before merging and resolve the complaints before merge. In this case, a release note label is missing (loosely, major = feature; minor = user-facing; bug = faulty behaviour; ci = testing-only; misc = dev-facing), and the title of the PR should be updated to reflect the actual change being made here. Note that this will end up in the release notes for the v1.11 final release, so it can be helpful to form a proper (short) sentence about what this PR achieves so folks can scan through and get some sense of the impact later on.

From a glance, this seems like it should be `release-note/minor` with a title something like `Support BGP route withdrawal`.

There's also a few others that need similar treatment to fix up the release notes for v1.11, see [this search](https://github.com/cilium/cilium/pulls?q=is%3Apr+author%3Aldelossa+is%3Aclosed).

**Comment by ldelossa** [comment]
@joestringer talked to you a bit out of band, but this is a topic branch merge, not a merge into master. PodCidr branch is still open and under review.

**Pull request merged**
